br/pkg/lightning/backend/external/iter.go

```diff
@@ -619,24 +646,80 @@ func newMergePropBaseIter(
 		// to support this.
 		limit = multiStat.MaxOverlappingNum + 1
 	}
+	limit = min(limit, int64(len(multiStat.Filenames)))
+
+	// we are rely on the caller have reduced the overall overlapping to less than
+	// MergeSortOverlapThreshold for []MultipleFilesStat. And we are going to open
+	// about 8000 connection to read files.
+	preOpenLimit := limit * 8
+	preOpenLimit = min(preOpenLimit, int64(len(multiStat.Filenames)))
+	preOpenCh := make(chan chan readerAndError, preOpenLimit-limit)
+	closeCh := make(chan struct{})
+	go func() {
+		defer close(preOpenCh)
+		// newLimitSizeMergeIter will open #limit readers at the beginning, and for rest
+		// readers we open them in advance to reduce block when we need to open them.
+		for i := int(limit); i < len(multiStat.Filenames); i++ {
+			filePair := multiStat.Filenames[i]
+			path := filePair[1]
+			asyncTask := make(chan readerAndError, 1)
+			go func() {
+				defer close(asyncTask)
+				rd, err := newStatsReader(ctx, exStorage, path, 500*1024)
+				select {
+				case <-closeCh:
+					return
+				case asyncTask <- readerAndError{r: &statReaderProxy{p: path, r: rd}, err: err}:
+				}
+			}()
+			select {
+			case <-closeCh:
```

Comment:
drain preOpenCh and all it's item before return?

Reply:
https://github.com/pingcap/tidb/pull/49832/commits/25b9e94e31eb29cc1ba3fb0c647239b00fe8d782